src/file_scanner/arrow_multi_file_info.cpp

```diff
@@ -0,0 +1,244 @@
+                                           shared_ptr<BaseFileReader>& reader,
+                                           GlobalTableFunctionState& gstate_p,
+                                           LocalTableFunctionState& lstate_p) {
+  auto& gstate = gstate_p.Cast<ArrowFileGlobalState>();
+  auto& lstate = lstate_p.Cast<ArrowFileLocalState>();
+  if (gstate.files.find(reader->file_list_idx.GetIndex()) != gstate.files.end()) {
+    return false;
+  }
+  gstate.files.insert(reader->file_list_idx.GetIndex());
+
+  lstate.file_scan = shared_ptr_cast<BaseFileReader, ArrowFileScan>(reader);
+  lstate.local_arrow_function_data = make_uniq<ArrowScanFunctionData>(
+      &FileIPCStreamFactory::Produce,
+      reinterpret_cast<uintptr_t>(lstate.file_scan->factory.get()));
+  lstate.local_arrow_function_data->schema_root = lstate.file_scan->schema_root;
+  lstate.local_arrow_function_data->arrow_table = lstate.file_scan->arrow_table_type;
+  if (!reader->column_indexes.empty()) {
+    lstate.init_input = make_uniq<TableFunctionInitInput>(
+        *lstate.local_arrow_function_data, reader->column_indexes,
+        gstate.global_state.projection_ids, reader->filters);
+  } else {
+    lstate.init_input = make_uniq<TableFunctionInitInput>(
+        *lstate.local_arrow_function_data, gstate.global_state.column_indexes,
+        gstate.global_state.projection_ids, reader->filters);
+  }
+  lstate.local_arrow_global_state =
+      ArrowTableFunction::ArrowScanInitGlobal(context, *lstate.init_input);
+  lstate.local_arrow_local_state =
+      ArrowTableFunction::ArrowScanInitLocal(lstate.execution_context, *lstate.init_input,
+                                             lstate.local_arrow_global_state.get());
+  lstate.table_function_input = make_uniq<TableFunctionInput>(
+      lstate.local_arrow_function_data.get(), lstate.local_arrow_local_state.get(),
+      lstate.local_arrow_global_state.get());
+  return true;
+}
+
+void ArrowMultiFileInfo::Scan(ClientContext& context, BaseFileReader& reader,
+                              GlobalTableFunctionState& global_state,
+                              LocalTableFunctionState& local_state, DataChunk& chunk) {
+  auto& lstate = local_state.Cast<ArrowFileLocalState>();
+  ArrowTableFunction::ArrowScanFunction(context, *lstate.table_function_input, chunk);
+}
+
+void ArrowMultiFileInfo::FinishFile(ClientContext& context,
+                                    GlobalTableFunctionState& global_state,
+                                    BaseFileReader& reader) {}
+
+void ArrowMultiFileInfo::FinishReading(ClientContext& context,
+                                       GlobalTableFunctionState& global_state,
+                                       LocalTableFunctionState& local_state) {}
+
+unique_ptr<NodeStatistics> ArrowMultiFileInfo::GetCardinality(
+    const MultiFileBindData& bind_data, idx_t file_count) {
+  return make_uniq<NodeStatistics>();
```

Comment:
This should at least have a `//! FIXME: keep track of the cardinality of the file(s) to allow the optimizer to work more efficiently`

What parquet does is this:
```
return make_uniq<NodeStatistics>(MaxValue(bind_data.initial_file_cardinality, (idx_t)1) * file_count);
```

Where the `initial_file_cardinality` comes from the "initial_reader" when not using `union_by_name` (not sure where it comes from in the `union_by_name` case tbh)